File: packages/host/src/workspace.ts
import { mkdirSync, writeFileSync, readFileSync, existsSync } from "node:fs";
import { TOOL_CATALOG, type CurrentPage, type TabsSnapshot } from "../../shared/src/protocol.ts";
import {
  AGENTS_MD_PATH,
  CLAUDE_MD_PATH,
  BROWSER_DIR,
  COMMANDS_DIR,
  CURRENT_PAGE_PATH,
  INTERACTIVE_PATH,
  PASTED_DIR,
  RESULTS_DIR,
  SCREENSHOTS_DIR,
  SESSION_PATH,
  SIDEBAR_HOME,
  SNAPSHOT_PATH,
  TABS_PATH,
  TOOLS_PATH,
  WORKSPACE_DIR,
} from "./paths.ts";

const AGENTS_MD = `# Browser page tools

You are chatting from a Chrome sidebar. The user keeps ONE workspace across every browser tab, and may switch among several chat sessions. Page files in this workspace are shared; the current conversation is only the session you are in.

Read \`browser/tools.json\` now. It lists every page method available in this session (read, act, vision). Prefer those methods over guessing from memory.

## Current page

When the user talks about "this page", "the current tab", or the site they are looking at, read these files first:

1. \`browser/current.json\` — tabId, url, title
2. **\`browser/interactive.md\`** — numbered interactive controls. This is how you fill forms and click. Do **not** start by guessing CSS selectors.
3. \`browser/snapshot.md\` — the same control list plus a readable text extract

The current tab also changes while you work. Re-read those files after navigation, after \`switchTab\`, or if the user says they changed pages.

## How to operate the page (read this)

Interactive controls look like:

\`\`\`
[1] textbox "Email" empty placeholder="you@acme.com" required
[2] textbox "Password" type=password empty
[3] combobox "Country" value="Select…" options=China | Japan
[4] checkbox "Remember me" unchecked
[5] button "Submit"
\`\`\`

Rules:

- Only lines with \`[index]\` are actionable. Indices are **1-based**.
- Prefer \`args.index\` from the latest list. Second choice: \`args.label\` or \`args.name\` (these resolve to the control, not the \`<label>\` node). CSS \`args.selector\` is a last resort.
- **Fill a whole form with one \`fillForm\`** whenever you have several fields. Do not invent a selector per box.
- After \`click\` / \`fill\` / \`fillForm\` / \`select\`, the result JSON includes a fresh \`interactive\` list. **Use those new indices.** Dropdowns, validation, and navigation change the numbering.
- Custom dropdowns: \`fill\` or \`select\` with the option text as \`args.value\`. If that fails, \`click\` the field, read the new list, \`click\` the option.
- Do not match inputs by visible \`innerText\` — empty textboxes have none. That is why \`interactive.md\` exists.
- File inputs cannot be filled. Ask the user to pick the file.
- Screenshots are for layout / confirmation, not the default way to find a form field.

### fillForm example

\`\`\`json
{"id":"cmd_form","method":"fillForm","args":{"fields":[{"label":"Email","value":"ada@example.com"},{"label":"Password","value":"secret"},{"label":"Country","value":"China"}]}}
\`\`\`

### click / fill by index

\`\`\`json
{"id":"cmd_01","method":"fill","args":{"index":1,"value":"ada@example.com"}}
{"id":"cmd_02","method":"click","args":{"index":5}}
\`\`\`

## Open tabs and windows

Read \`browser/tabs.json\` for every normal Chrome window and tab (\`tabId\`, \`windowId\`, \`index\`, \`title\`, \`url\`, \`active\`, \`pinned\`, \`restricted\`). It updates when tabs move. Call \`listTabs\` if you need the same list in a command result.

Do not invent tab IDs. To switch tabs, call \`switchTab\` with \`args.tabId\` from that file — it also focuses the tab's window. To open a site without replacing the current page, call \`openTab\` with \`args.url\` (http(s) only). To pull one or more tabs into their own window, call \`moveTabsToWindow\` with \`args.tabIds\`. Pass \`args.windowId\` to move them into an existing window instead of creating one.

\`restricted: true\` means chrome://, chrome-extension://, or the Web Store. You may \`switchTab\` to those, but do not run page read / act / screenshot on them.

If the user message includes \`[Picked page elements]\`, those CSS selectors were chosen by the user in the sidebar picker. Inspect or operate on that exact node with page tools and \`args.selector\`. Do not treat those lines as file paths.

If the user message includes \`[Mentioned tabs]\`, the user @-mentioned those browser tabs in the composer. Inline \`@Title\` names match the titles in that block. Use \`switchTab\` with the given tabId if it still appears in \`browser/tabs.json\`; otherwise \`openTab\` the URL (http(s) only).

If the user message includes \`[Mentioned attachments]\`, those are files / folders / images the user @-mentioned. Read the local paths. If it includes \`[Mentioned page elements]\`, use page tools with \`args.selector\` set to that CSS selector.

## Calling page methods

Write a JSON file to \`browser/commands/<id>.json\`, then read \`browser/results/<id>.json\`. If the result is not there yet, wait a moment and read again.

Find elements with \`args.index\` (from \`interactive.md\`), \`args.label\` / \`args.name\`, \`args.selector\` (CSS), \`args.text\` (label then visible text), and optional \`args.nth\` (0-based among matches).

### Read

- \`getInteractive\` — refresh the numbered control list (same as \`browser/interactive.md\`)
- \`getMeta\` — url, title, description
- \`getReadable\` — main text
- \`getSelection\` — highlighted text
- \`getLinks\` — same-origin links
- \`getOutline\` — h1–h3
- \`queryText\` — one node's text
- \`queryAll\` — matching node summaries; no locator = current interactive list
- \`getAttribute\` — requires \`args.attribute\`
- \`getValue\` — current value + label of a field
- \`exists\` — whether a match exists
- \`listTabs\` — all normal windows and tabs (same shape as \`browser/tabs.json\`)

### Act

- \`click\` / \`dblclick\` / \`hover\` / \`focus\`
- \`fillForm\` — many fields at once (\`args.fields\`)
- \`fill\` — set value (\`args.value\`)
- \`type\` — append \`args.text\`
- \`clear\`
- \`select\` — native \`<select>\` or custom combobox; \`args.value\` is option value or visible text
- \`check\` — checkbox/radio/switch, optional \`args.checked\`
- \`press\` — \`args.key\` such as \`Enter\` or \`Escape\`
- \`scroll\` — element if index/selector/text, else window by \`args.x\` / \`args.y\`
- \`scrollIntoView\`
- \`waitFor\` — poll until the element exists (\`args.timeoutMs\`, max 20000)
- \`navigate\` — \`args.url\`, http(s) only
- \`goBack\` / \`goForward\` / \`reload\`
- \`switchTab\` — \`args.tabId\`
- \`openTab\` — \`args.url\` (http(s)), optional \`args.windowId\`
- \`moveTabsToWindow\` — \`args.tabIds\`, optional \`args.windowId\`

### Vision

Use screenshots when the control list is not enough to understand layout, pick a target, or confirm what changed.

- \`screenshot\` — visible viewport JPEG; optional \`x,y,width,height\` in CSS pixels
- \`screenshotElement\` — scroll an element into view and crop it (\`index\`, \`selector\`, or \`label\`)

The result JSON has \`data.path\` (absolute file). **Read that JPEG** to inspect the page visually. Do not expect base64 in the JSON.

Do not invent other methods. Do not try to run arbitrary JavaScript. After acting, re-read \`browser/current.json\`, \`browser/interactive.md\`, and \`browser/tabs.json\` if the page or tab set may have changed.
`;

export function ensureWorkspace(): void {
  mkdirSync(COMMANDS_DIR, { recursive: true });
  mkdirSync(RESULTS_DIR, { recursive: true });
  mkdirSync(SCREENSHOTS_DIR, { recursive: true });
  mkdirSync(PASTED_DIR, { recursive: true });
  mkdirSync(SIDEBAR_HOME, { recursive: true });
  writeFileSync(AGENTS_MD_PATH, AGENTS_MD);
  writeFileSync(CLAUDE_MD_PATH, AGENTS_MD);
  writeFileSync(
    TOOLS_PATH,
    `${JSON.stringify(
      {
        version: 4,
        transport: "workspace-files",
        commandsDir: "browser/commands",
        resultsDir: "browser/results",
        screenshotsDir: "browser/screenshots",
        methods: TOOL_CATALOG,
      },
      null,
      2,
    )}\n`,
  );
}

export function writeCurrentPage(page: CurrentPage): void {
  mkdirSync(BROWSER_DIR, { recursive: true });
  const compact = {
    tabId: page.tabId,
    url: page.url,
    title: page.title,
    updatedAt: page.updatedAt,
  };
  writeFileSync(CURRENT_PAGE_PATH, `${JSON.stringify(compact, null, 2)}\n`);
  const interactive = page.interactive?.trim() || "_No interactive controls indexed._";
  const readable = page.readable?.trim() || "_No readable extract available._";
  writeFileSync(INTERACTIVE_PATH, `# ${page.title}\n\n${page.url}\n\n${interactive}\n`);
  writeFileSync(
    SNAPSHOT_PATH,
    `# ${page.title}\n\n${page.url}\n\n## Interactive controls\n\nSee \`browser/interactive.md\`. Use \`args.index\` (1-based) or \`fillForm\`.\n\n${interactive}\n\n## Readable text\n\n${readable}\n`,
  );
}

export function writeTabsSnapshot(snapshot: TabsSnapshot): void {
  mkdirSync(BROWSER_DIR, { recursive: true });
  writeFileSync(TABS_PATH, `${JSON.stringify(snapshot, null, 2)}\n`);
}

export function readSessionId(): string | undefined {
  if (!existsSync(SESSION_PATH)) return undefined;
  try {
    const data = JSON.parse(readFileSync(SESSION_PATH, "utf8")) as { sessionId?: string };
    return data.sessionId;
  } catch {
    return undefined;
  }
}

export function writeSessionId(sessionId: string): void {
  mkdirSync(SIDEBAR_HOME, { recursive: true });
  writeFileSync(SESSION_PATH, `${JSON.stringify({ sessionId }, null, 2)}\n`);
}

export { WORKSPACE_DIR };

File: internal/workspace/workspace.go
package workspace

import (
	"encoding/json"
	"os"
	"path/filepath"
	"strings"

	_ "embed"

	"github.com/parksben/opensider/internal/paths"
	"github.com/parksben/opensider/internal/protocol"
)

//go:embed agents.md
var agentsMD string

func Ensure() {
	_ = os.MkdirAll(paths.CommandsDir(), 0o755)
	_ = os.MkdirAll(paths.ResultsDir(), 0o755)
	_ = os.MkdirAll(paths.ScreenshotsDir(), 0o755)
	_ = os.MkdirAll(paths.PastedDir(), 0o755)
	_ = os.MkdirAll(paths.OutputsDir(), 0o755)
	_ = os.MkdirAll(paths.SidebarHome(), 0o755)
	_ = os.WriteFile(paths.AgentsMDPath(), []byte(agentsMD), 0o644)
	_ = os.WriteFile(paths.ClaudeMDPath(), []byte(agentsMD), 0o644)
	tools := map[string]any{
		"version":        12,
		"transport":      "workspace-files",
		"commandsDir":    "browser/commands",
		"resultsDir":     "browser/results",
		"screenshotsDir": "browser/screenshots",
		"outputsDir":     "outputs",
		// While the side panel is open the tab the Agent works with is forced to look
		// visible and focused to the page (see agents.md). Static description: the live
		// state is per tab and per moment.
		"pageActivity": map[string]any{
			"forcedWhilePanelOpen": true,
			"stealsWindowFocus":    false,
			"note":                 "The tab you operate reads document.visibilityState === \"visible\" and hasFocus() === true, and does not receive visibilitychange/blur/pagehide/freeze, while the side panel is open.",
		},
		// Quiet defaults + tab control (see agents.md «Tab control»): static description only.
		"tabControl": map[string]any{
			"quietByDefault": true,
			"note": "openTab opens in the background next to your working tab and becomes yours; no command changes the user's active tab or window focus, except switchTab (show) and screenshots (a brief flip that is restored).",
		},
		"methods": protocol.ToolCatalog,
	}
	raw, _ := json.MarshalIndent(tools, "", "  ")
	_ = os.WriteFile(paths.ToolsPath(), append(raw, '\n'), 0o644)
}

func WriteCurrentPage(page protocol.CurrentPage) {
	_ = os.MkdirAll(paths.BrowserDir(), 0o755)
	compact := map[string]any{
		"tabId":     page.TabID,
		"url":       page.URL,
		"title":     page.Title,
		"updatedAt": page.UpdatedAt,
	}
	if page.Target != nil {
		compact["target"] = page.Target
	}
	raw, _ := json.MarshalIndent(compact, "", "  ")
	_ = os.WriteFile(paths.CurrentPagePath(), append(raw, '\n'), 0o644)
	interactive := page.Interactive
	if trimSpace(interactive) == "" {
		interactive = "_No interactive controls indexed._"
	}
	readable := page.Readable
	if trimSpace(readable) == "" {
		readable = "_No readable extract available._"
	}
	_ = os.WriteFile(paths.InteractivePath(), []byte("# "+page.Title+"\n\n"+page.URL+"\n\n"+interactive+"\n"), 0o644)
	snapshot := "# " + page.Title + "\n\n" + page.URL + "\n\n## Interactive controls\n\nSee `browser/interactive.md`. Use `args.index` (1-based) or `fillForm`.\n\n" + interactive + "\n\n## Readable text\n\n" + readable + "\n"
	_ = os.WriteFile(paths.SnapshotPath(), []byte(snapshot), 0o644)
}

func WriteTabsSnapshot(snapshot protocol.TabsSnapshot) {
	_ = os.MkdirAll(paths.BrowserDir(), 0o755)
	raw, _ := json.MarshalIndent(snapshot, "", "  ")
	_ = os.WriteFile(paths.TabsPath(), append(raw, '\n'), 0o644)
}

func WriteSessionID(sessionID string) {
	_ = os.MkdirAll(paths.SidebarHome(), 0o755)
	raw, _ := json.MarshalIndent(map[string]string{"sessionId": sessionID}, "", "  ")
	_ = os.WriteFile(paths.SessionPath(), append(raw, '\n'), 0o644)
}

func SavePastedJPEG(imageBase64 string, suggestedName string) (protocol.AttachmentItem, error) {
	const maxBase64 = 800_000
	if imageBase64 == "" || len(imageBase64) > maxBase64 {
		return protocol.AttachmentItem{}, errPastedTooLarge
	}
	_ = os.MkdirAll(paths.PastedDir(), 0o755)
	stamp := isoStamp()
	raw := filepath.Base(suggestedName)
	if raw == "" || raw == "." {
		raw = "paste-" + stamp + ".jpg"
	}
	safe := sanitizeName(raw)
	if safe == "" {
		safe = "paste-" + stamp + ".jpg"
	}
	name := safe
	if !hasJPEGExt(name) {
		name += ".jpg"
	}
	p := filepath.Join(paths.PastedDir(), name)
	if _, err := os.Stat(p); err == nil {
		p = filepath.Join(paths.PastedDir(), "paste-"+stamp+".jpg")
	}
	data, err := decodeBase64(imageBase64)
	if err != nil {
		return protocol.AttachmentItem{}, err
	}
	if err := os.WriteFile(p, data, 0o644); err != nil {
		return protocol.AttachmentItem{}, err
	}
	return protocol.AttachmentItem{Path: p, Name: filepath.Base(p), Kind: protocol.KindImage}, nil
}

var errPastedTooLarge = pastedError("pasted image is empty or too large")

// SaveUploaded 把用户拖进侧栏的一个文件写到 `browser/uploads/` 下，并回它对应的附件项。
//
// Chrome 不给拖入文件的本机路径（DataTransfer 里只有 File 对象），所以侧栏把字节送过来、
// 这里落盘，之后一切都按普通路径附件走（Agent 读的是工作区里的这份副本）。
//
//   - 单个文件：`name` 是文件名，`dir` 为空；
//   - 文件夹里的文件：`dir` 是拖进来的文件夹名，`name` 是该文件夹内的相对路径，
//     例如 `dir=proj` + `name=src/a.ts` 写成 `uploads/proj/src/a.ts`。
//
// `dir` / `name` 逐段 sanitize，拒绝空段、`.`、`..` 与绝对路径；同名文件加时间戳后缀，
// 不覆盖之前的。
func SaveUploaded(name, dir, encoded string) ([]protocol.AttachmentItem, error) {
	const maxBase64 = 800_000
	if encoded == "" || len(encoded) > maxBase64 {
		return nil, errUploadTooLarge
	}
	rel, err := uploadRelPath(name, dir)
	if err != nil {
		return nil, err
	}
	root := paths.UploadsDir()
	target := filepath.Join(root, rel)
	// 双重保险：拼出来的路径必须真的落在 uploads/ 里。
	if !strings.HasPrefix(target, root+string(os.PathSeparator)) {
		return nil, errUploadPath
	}
	created := false
	if dir != "" {
		folder := filepath.Join(root, sanitizeSegment(dir))
		if _, statErr := os.Stat(folder); statErr != nil {
			created = true
		}
	}
	if mkErr := os.MkdirAll(filepath.Dir(target), 0o755); mkErr != nil {
		return nil, mkErr
	}
	if _, statErr := os.Stat(target); statErr == nil {
		ext := filepath.Ext(target)
		stem := strings.TrimSuffix(target, ext)
		target = stem + "-" + isoStamp() + ext
	}
	data, err := decodeBase64(encoded)
	if err != nil {
		return nil, err
	}
	if err := os.WriteFile(target, data, 0o644); err != nil {
		return nil, err
	}
	items := make([]protocol.AttachmentItem, 0, 2)
	// 文件夹只在这次真的建出来时带一条：侧栏按 path 去重，多报也无害，但少报会让
	// 用户看不到那个文件夹附件。
	if dir != "" && created {
		folder := filepath.Join(root, sanitizeSegment(dir))
		items = append(items, protocol.AttachmentItem{
			Path: folder,
			Name: filepath.Base(folder),
			Kind: protocol.KindFolder,
		})
	}
	items = append(items, protocol.AttachmentItem{
		Path: target,
		Name: filepath.Base(target),
		Kind: protocol.KindFile,
	})
	return items, nil
}

// uploadRelPath 把 (name, dir) 变成 uploads/ 下的相对路径；任何可疑段都直接报错。
func uploadRelPath(name, dir string) (string, error) {
	segments := []string{}
	if dir != "" {
		// 文件夹名只是拖进来那一层的名字，出现路径语法说明输入不可能是我们发出去的。
		if dir == "." || dir == ".." || strings.ContainsAny(dir, `/\`) || filepath.IsAbs(dir) {
			return "", errUploadPath
		}
		segments = append(segments, sanitizeSegment(dir))
	}
	for _, part := range strings.Split(strings.ReplaceAll(name, "\\", "/"), "/") {
		if part == "" || part == "." {
			continue
		}
		if part == ".." {
			return "", errUploadPath
		}
		segments = append(segments, sanitizeSegment(part))
	}
	if len(segments) == 0 || filepath.IsAbs(name) {
		return "", errUploadPath
	}
	return filepath.Join(segments...), nil
}

func sanitizeSegment(part string) string {
	safe := sanitizeName(part)
	if safe == "" || safe == "." || safe == ".." {
		return "upload-" + isoStamp()
	}
	return safe
}

var (
	errUploadTooLarge = pastedError("dropped file is empty or too large")
	errUploadPath     = pastedError("dropped file has an unusable name")
)

type pastedError string

func (e pastedError) Error() string { return string(e) }
